[EuiFilterGroup] Add `compressed` prop

miukimiu opened the pull request:
### Summary

This PR adds a `compressed` prop to **EuiFilterGroup** and reduces the size of the **EuiFilterButton** notification badge. It also closes #5695.

As mentioned in #5695, one of the ways to solve this issue was to inherit the **EuiButton** prop `size` in **EuiFilterButton**.  But IMO supporting different sizes could cause some inconsistencies in usage. Consumers could pass one size to one button and other sizes to others. Also passing a size prop to multiple components seems very a repetitive task when we can achieve the same goal by just passing one prop in the parent component.

For this reason, I decided to add a `compressed` prop in the **EuiFilterButton** parent component **EuiFilterGroup**.

Also, the **EuiFilterGroup** is part of our **forms** section and I think it makes sense to also support the `compressed` prop like other form components.

<img width="579" alt="EuiFilterGroup-compressed@2x" src="https://user-images.githubusercontent.com/2750668/158234835-3f13f9ba-6ebf-4086-ab02-1d9ac574daaf.png">


### Checklist

- [x] Checked in both **light and dark** modes
- ~[ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [x] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**
- [x] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples
- [x] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart
- [x] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

miukimiu commented on src/components/filter_group/filter_button.tsx:
```diff
@@ -137,6 +143,7 @@ export const EuiFilterButton: FunctionComponent<EuiFilterButtonProps> = ({
   );
 
   return (
+    // @ts-ignore type error
```
I don't know how to solve this TS error. It happens when I:

```
Omit<
  EuiButtonEmptyProps,
  'flush' | 'size'
>
```

 in `type EuiFilterButtonProps`

thompsongl replied:
Use `DistributiveOmit` instead. Import it from `common.ts`.

cchaos commented on src/themes/amsterdam/overrides/_filter_group.scss:
```diff
@@ -5,6 +5,15 @@
+    height: $euiButtonHeightSmall;
+  }
+}
+
```
Nit: Extra space
```suggestion
```

cchaos commented on src/components/filter_group/filter_group.tsx:
```diff
@@ -17,18 +17,24 @@ export type EuiFilterGroupProps = HTMLAttributes<HTMLDivElement> &
      */
     fullWidth?: boolean;
+    /**
+     *  When `true` creates a shorter height filter group
```
```suggestion
     *  When `true` creates a shorter height filter group matching that of `compressed` form controls
```

cchaos commented on src/themes/amsterdam/overrides/_filter_group.scss:
```diff
@@ -5,6 +5,15 @@
+  border-radius: $euiFormControlCompressedBorderRadius;
+
+  & .euiFilterButton {
+    height: $euiButtonHeightSmall;
```
Because you're prop name is actually trying to be consistent with form sizing, I'd use the form specific variables for everything.

```suggestion
    height: $euiFormControlCompressedHeight;
```

cchaos commented on src/components/filter_group/filter_button.tsx:
```diff
@@ -11,11 +11,17 @@ import classNames from 'classnames';
 import { useInnerText } from '../inner_text';
 
-export type EuiFilterButtonProps = EuiButtonEmptyProps & {
+export type EuiFilterButtonProps = Omit<
+  EuiButtonEmptyProps,
+  'flush' | 'size'
+> & {
```
On thing that I've noticed with our Props tables in the docs is that it spits out the props in the order that we supply them here. So in this case, the custom filter button props will display **after** all the **EuiButtonEmptyProps**. Can I make a suggestion and put this **after** the filter button props?

cchaos replied:
Also removing these props may be considered a breaking change in case users are in fact using this props?

cchaos reviewed (commented):
**Suggestion**: How about also decreasing the EuiNotificationBadge size to `s` when compressed?

cchaos commented on src/themes/amsterdam/overrides/_filter_group.scss:
```diff
@@ -5,6 +5,21 @@
+  }
+
+  // Override the size to match the default `.euiNotificationBadge` size
+  .euiFilterButton__notification {
+    line-height: $euiSize;
+    height: $euiSize;
+    min-width: $euiSize;
+  }
```
Oh right, because you're not applying a size to the filter button itself, there's no way to set `size="s"` directly on the notification badge.... 🤔 

This is kind of a scary override/customization to maintain.

What if it's just always `size="s"`? And remove any current overrides to the style. It feels a bit large to begin with.

miukimiu replied:
One thing that I did first was to remove the `size="m"` and when the `.euiFilterGroup:not(.euiFilterGroup--compressed) ` I was extending the `.euiNotificationBadge--medium`. But I noticed that we don't use Sass extends regularly so I decided to override. 😬 

But just leave the default `size="s"` looks good to me.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5717/

miukimiu commented:
Thanks, @cchaos, and @thompsongl. I went through the review and fixed everything. 

@cchaos I also updated the Figma component and added a compressed variant. I already published the changes. The notification badges there were already using the size small.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5717/

thompsongl reviewed (approved):
Code changes LGTM!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5717/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5717/
